src/main/java/week4/unionFind/IslandCountUF.java

```diff
@@ -0,0 +1,74 @@
+package week4.unionFind;
+
+
+import java.util.Arrays;
+import week4.IslandCountI;
+
+public class IslandCountUF implements IslandCountI {
+
+  /*
+   * We go through each cell, if it is the top-left corner of an island,
+   * we increase the count if it is part of an island, we merge it in the UF DS.
+   * We do a second pass through the map to correct over-estimated count
+   * (they are connected through right or bottom tile)
+   *
+   * Time Complexity: O(nm) - pass through the map twice
+   * Space Complexity: O(nm) - the size of the map = size of the UF DS
+   */
+  public int countIslands(boolean[][] map) {
+
+    if (map == null || map.length == 0) {
+      return 0;
+    }
+
+    int count = 0;
+    int rows = map.length;
+    int cols = map[0].length;
+    int[] uf = new int[rows * cols];
+
+    Arrays.setAll(uf, i -> i);
+
+    for (int r = 0; r < rows; r++) {
+      for (int c = 0; c < cols; c++) {
+        if (map[r][c]) {
+          if (r > 0 && map[r - 1][c]) {
+            merge(uf, getIndex(r, c, rows), getIndex(r - 1, c, rows));
+          } else if (c > 0 && map[r][c - 1]) {
```

Comment:
yep - found the counter example it shows up in the bottom right corner so apparently I did not test for the corners... fix still to be done